File: src/media/video/mpp_loader.c
#include "mpp_loader.h"

#include <dlfcn.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef uint32_t (*AbiVersion)(void);
typedef GoMppDecoder* (*CreateDecoder)(int max_width, int max_height);
typedef int (*SubmitDecoder)(GoMppDecoder* decoder, const uint8_t* data, size_t length);
typedef int (*ReceiveDecoder)(GoMppDecoder* decoder, GoMppFrame* frame);
typedef void (*ReleaseFrame)(GoMppDecoder* decoder, GoMppFrame* frame);
typedef int (*ResetDecoder)(GoMppDecoder* decoder);
typedef const char* (*DecoderError)(GoMppDecoder* decoder);
typedef void (*DestroyDecoder)(GoMppDecoder* decoder);

struct GoMppLibrary {
    void* handle;
    GoMppDecoder* decoder;
    SubmitDecoder submit;
    ReceiveDecoder receive;
    ReleaseFrame release_frame;
    ResetDecoder reset;
    DecoderError decoder_error;
    DestroyDecoder destroy;
    char path[512];
    char error[256];
};

static void set_loader_error(GoMppLibrary* library, const char* message) {
    snprintf(library->error, sizeof(library->error), "%s", message ? message : "unknown error");
}

static void* load_symbol(GoMppLibrary* library, const char* name) {
    dlerror();
    void* symbol = dlsym(library->handle, name);
    const char* error = dlerror();
    if (error || !symbol)
        set_loader_error(library, error ? error : "MPP plugin symbol is missing");
    return symbol;
}

GoMppLibrary* go_mpp_library_open(int max_width, int max_height) {
    GoMppLibrary* library = calloc(1, sizeof(*library));
    if (!library)
        return NULL;

    const char* path = getenv("GREENOVERCAST_MPP_LIBRARY");
    if (!path || !path[0])
        path = "libgreenovercast-mpp.so";
    if (snprintf(library->path, sizeof(library->path), "%s", path) >= (int)sizeof(library->path)) {
        set_loader_error(library, "MPP plugin path is too long");
        return library;
    }

    int flags = RTLD_NOW | RTLD_LOCAL;
#ifdef RTLD_NODELETE
    /* Older Rockchip MPP builds can run process-lifetime cleanup after decoder teardown. */
    flags |= RTLD_NODELETE;
#endif
    library->handle = dlopen(path, flags);
    if (!library->handle) {
        set_loader_error(library, dlerror());
        return library;
    }

    AbiVersion abi_version = (AbiVersion)load_symbol(library, "go_mpp_decoder_abi_version");
    if (!abi_version)
        goto fail;
    uint32_t version = abi_version();
    if (version != GO_MPP_DECODER_ABI_VERSION) {
        snprintf(library->error, sizeof(library->error),
                 "MPP plugin ABI %u is incompatible with required ABI %u", version,
                 GO_MPP_DECODER_ABI_VERSION);
        goto fail;
    }

    CreateDecoder create = (CreateDecoder)load_symbol(library, "go_mpp_decoder_create");
    library->submit = (SubmitDecoder)load_symbol(library, "go_mpp_decoder_submit");
    library->receive = (ReceiveDecoder)load_symbol(library, "go_mpp_decoder_receive");
    library->release_frame = (ReleaseFrame)load_symbol(library, "go_mpp_decoder_release_frame");
    library->reset = (ResetDecoder)load_symbol(library, "go_mpp_decoder_reset");
    library->decoder_error = (DecoderError)load_symbol(library, "go_mpp_decoder_last_error");
    library->destroy = (DestroyDecoder)load_symbol(library, "go_mpp_decoder_destroy");
    if (!create || !library->submit || !library->receive || !library->release_frame ||
        !library->reset || !library->decoder_error || !library->destroy)
        goto fail;

    library->decoder = create(max_width, max_height);
    if (!library->decoder) {
        set_loader_error(library, library->decoder_error(NULL));
        return library;
    }
    return library;

fail:
    dlclose(library->handle);
    library->handle = NULL;
    return library;
}

int go_mpp_library_ready(const GoMppLibrary* library) {
    return library && library->decoder && library->submit && library->receive &&
           library->release_frame && library->reset && library->decoder_error && library->destroy;
}

int go_mpp_library_submit(GoMppLibrary* library, const uint8_t* data, size_t length) {
    if (!go_mpp_library_ready(library))
        return -1;
    int result = library->submit(library->decoder, data, length);
    if (result < 0)
        set_loader_error(library, library->decoder_error(library->decoder));
    return result;
}

int go_mpp_library_receive(GoMppLibrary* library, GoMppFrame* frame) {
    if (!go_mpp_library_ready(library))
        return -1;
    int result = library->receive(library->decoder, frame);
    if (result < 0)
        set_loader_error(library, library->decoder_error(library->decoder));
    return result;
}

void go_mpp_library_release_frame(GoMppLibrary* library, GoMppFrame* frame) {
    if (go_mpp_library_ready(library) && frame)
        library->release_frame(library->decoder, frame);
}

int go_mpp_library_reset(GoMppLibrary* library) {
    if (!go_mpp_library_ready(library))
        return -1;
    int result = library->reset(library->decoder);
    if (result < 0)
        set_loader_error(library, library->decoder_error(library->decoder));
    return result;
}

const char* go_mpp_library_error(const GoMppLibrary* library) {
    return library && library->error[0] ? library->error : "MPP decoder unavailable";
}

const char* go_mpp_library_path(const GoMppLibrary* library) {
    return library && library->path[0] ? library->path : "libgreenovercast-mpp.so";
}

void go_mpp_library_close(GoMppLibrary* library) {
    if (!library)
        return;
    if (library->decoder && library->destroy)
        library->destroy(library->decoder);
    if (library->handle)
        dlclose(library->handle);
    free(library);
}

File: src/media/video/mpp_loader.zig
const std = @import("std");

const c = @cImport({
    @cInclude("dlfcn.h");
    @cInclude("mpp_decoder.h");
});

const AbiVersion = *const fn () callconv(.c) u32;
const CreateDecoder = *const fn (c_int, c_int) callconv(.c) ?*c.GoMppDecoder;
const SubmitDecoder = *const fn (*c.GoMppDecoder, [*]const u8, usize) callconv(.c) c_int;
const ReceiveDecoder = *const fn (*c.GoMppDecoder, *c.GoMppFrame) callconv(.c) c_int;
const ReleaseFrame = *const fn (*c.GoMppDecoder, *c.GoMppFrame) callconv(.c) void;
const ResetDecoder = *const fn (*c.GoMppDecoder) callconv(.c) c_int;
const DecoderError = *const fn (?*c.GoMppDecoder) callconv(.c) [*:0]const u8;
const DestroyDecoder = *const fn (*c.GoMppDecoder) callconv(.c) void;

const Library = struct {
    handle: ?*anyopaque = null,
    decoder: ?*c.GoMppDecoder = null,
    submit: ?SubmitDecoder = null,
    receive: ?ReceiveDecoder = null,
    release_frame: ?ReleaseFrame = null,
    reset: ?ResetDecoder = null,
    decoder_error: ?DecoderError = null,
    destroy: ?DestroyDecoder = null,
    path: [512]u8 = [_]u8{0} ** 512,
    error_message: [256]u8 = [_]u8{0} ** 256,
};

fn copyZ(destination: []u8, message: []const u8) void {
    @memset(destination, 0);
    const length = @min(destination.len - 1, message.len);
    @memcpy(destination[0..length], message[0..length]);
}

fn setError(library: *Library, message: ?[*:0]const u8) void {
    copyZ(&library.error_message, if (message) |value| std.mem.span(value) else "unknown error");
}

fn loadSymbol(library: *Library, comptime T: type, name: [*:0]const u8) ?T {
    _ = c.dlerror();
    const symbol = c.dlsym(library.handle, name);
    const load_error = c.dlerror();
    if (load_error != null or symbol == null) {
        setError(
            library,
            if (load_error != null) @ptrCast(load_error) else "MPP plugin symbol is missing",
        );
        return null;
    }
    return @ptrCast(@alignCast(symbol));
}

fn ready(library: ?*const Library) bool {
    const value = library orelse return false;
    return value.decoder != null and value.submit != null and value.receive != null and
        value.release_frame != null and value.reset != null and value.decoder_error != null and
        value.destroy != null;
}

pub export fn go_mpp_library_open(max_width: c_int, max_height: c_int) ?*Library {
    const library = std.heap.c_allocator.create(Library) catch return null;
    library.* = .{};

    const path = std.posix.getenv("GREENOVERCAST_MPP_LIBRARY") orelse "libgreenovercast-mpp.so";
    if (path.len >= library.path.len) {
        copyZ(&library.error_message, "MPP plugin path is too long");
        return library;
    }
    copyZ(&library.path, path);

    var flags: c_int = c.RTLD_NOW | c.RTLD_LOCAL;
    // Older MPP builds can run process-lifetime cleanup after decoder teardown.
    if (@hasDecl(c, "RTLD_NODELETE")) flags |= c.RTLD_NODELETE;
    library.handle = c.dlopen(@ptrCast(&library.path), flags);
    if (library.handle == null) {
        const load_error = c.dlerror();
        setError(library, if (load_error != null) @ptrCast(load_error) else null);
        return library;
    }

    const abi_version = loadSymbol(library, AbiVersion, "go_mpp_decoder_abi_version") orelse {
        _ = c.dlclose(library.handle);
        library.handle = null;
        return library;
    };
    const version = abi_version();
    if (version != c.GO_MPP_DECODER_ABI_VERSION) {
        _ = std.fmt.bufPrintZ(
            &library.error_message,
            "MPP plugin ABI {d} is incompatible with required ABI {d}",
            .{ version, c.GO_MPP_DECODER_ABI_VERSION },
        ) catch copyZ(&library.error_message, "MPP plugin ABI is incompatible");
        _ = c.dlclose(library.handle);
        library.handle = null;
        return library;
    }

    const create = loadSymbol(library, CreateDecoder, "go_mpp_decoder_create");
    library.submit = loadSymbol(library, SubmitDecoder, "go_mpp_decoder_submit");
    library.receive = loadSymbol(library, ReceiveDecoder, "go_mpp_decoder_receive");
    library.release_frame = loadSymbol(library, ReleaseFrame, "go_mpp_decoder_release_frame");
    library.reset = loadSymbol(library, ResetDecoder, "go_mpp_decoder_reset");
    library.decoder_error = loadSymbol(library, DecoderError, "go_mpp_decoder_last_error");
    library.destroy = loadSymbol(library, DestroyDecoder, "go_mpp_decoder_destroy");
    if (create == null or library.submit == null or library.receive == null or
        library.release_frame == null or library.reset == null or library.decoder_error == null or
        library.destroy == null)
    {
        _ = c.dlclose(library.handle);
        library.handle = null;
        return library;
    }

    library.decoder = create.?(max_width, max_height);
    if (library.decoder == null) setError(library, library.decoder_error.?(null));
    return library;
}

pub export fn go_mpp_library_ready(library: ?*const Library) c_int {
    return @intFromBool(ready(library));
}

pub export fn go_mpp_library_submit(
    library_pointer: ?*Library,
    data: ?[*]const u8,
    length: usize,
) c_int {
    const library = library_pointer orelse return -1;
    const bytes = data orelse return -1;
    if (!ready(library)) return -1;
    const result = library.submit.?(library.decoder.?, bytes, length);
    if (result < 0) setError(library, library.decoder_error.?(library.decoder));
    return result;
}

pub export fn go_mpp_library_receive(library_pointer: ?*Library, frame: ?*c.GoMppFrame) c_int {
    const library = library_pointer orelse return -1;
    const output = frame orelse return -1;
    if (!ready(library)) return -1;
    const result = library.receive.?(library.decoder.?, output);
    if (result < 0) setError(library, library.decoder_error.?(library.decoder));
    return result;
}

pub export fn go_mpp_library_release_frame(
    library_pointer: ?*Library,
    frame: ?*c.GoMppFrame,
) void {
    const library = library_pointer orelse return;
    const output = frame orelse return;
    if (ready(library)) library.release_frame.?(library.decoder.?, output);
}

pub export fn go_mpp_library_reset(library_pointer: ?*Library) c_int {
    const library = library_pointer orelse return -1;
    if (!ready(library)) return -1;
    const result = library.reset.?(library.decoder.?);
    if (result < 0) setError(library, library.decoder_error.?(library.decoder));
    return result;
}

pub export fn go_mpp_library_error(library: ?*const Library) [*:0]const u8 {
    if (library) |value| {
        if (value.error_message[0] != 0) return @ptrCast(&value.error_message);
    }
    return "MPP decoder unavailable";
}

pub export fn go_mpp_library_path(library: ?*const Library) [*:0]const u8 {
    if (library) |value| {
        if (value.path[0] != 0) return @ptrCast(&value.path);
    }
    return "libgreenovercast-mpp.so";
}

pub export fn go_mpp_library_close(library_pointer: ?*Library) void {
    const library = library_pointer orelse return;
    if (library.decoder) |decoder| {
        if (library.destroy) |destroy| destroy(decoder);
    }
    if (library.handle != null) _ = c.dlclose(library.handle);
    std.heap.c_allocator.destroy(library);
}
